client/src/controllers/TabsController.ts

```diff
@@ -0,0 +1,317 @@
+/* eslint-disable no-shadow */
+import { Controller } from '@hotwired/stimulus';
+
+interface IndexedEventTarget extends EventTarget {
+  index: number;
+}
+
+interface TabLink extends HTMLAnchorElement {
+  index: number;
+}
+
+enum Keys {
+  Left = 'ArrowLeft',
+  Right = 'ArrowRight',
+  Home = 'Home',
+  End = 'End',
+}
+
+enum KeyWeight {
+  ArrowLeft = -1,
+  ArrowRight = 1,
+}
+
+/**
+ * @example - creating a simple tabs interface
+ * <div class="w-tabs" data-controller="w-tabs" data-action="popstate@window->w-tabs#loadHistory" data-w-tabs-active-value="" data-w-tabs-animate-class="animate-in">
+ *   <div class="w-tabs__list" role="tablist" data-w-tabs-target="list">
+ *       <a id="tab-label-tab-1" href="#tab-tab-1" class="w-tabs__tab " role="tab"
+ *          tabindex="-1" data-action="click->w-tabs#handleTabChange:prevent keydown->w-tabs#handleKeydown" data-w-tabs-target="label">
+ *         Tab 1
+ *       </a>
+ *       <a id="tab-label-tab-2" href="#tab-tab-2" class="w-tabs__tab " role="tab"
+ *          data-action="click->w-tabs#handleTabChange:prevent keydown->w-tabs#handleKeydown" data-w-tabs-target="label">
+ *         Tab 2
+ *       </a>
+ *   </div>
+ *
+ *   <div class="tab-content tab-content--comments-enabled">
+ *       <section id="tab-tab-1" class="w-tabs__panel " role="tabpanel" aria-labelledby="tab-label-tab-1" data-w-tabs-target="panel">
+ *           tab-1
+ *       </section>
+ *       <section id="tab-tab-2" class="w-tabs__panel " role="tabpanel" aria-labelledby="tab-label-tab-2" data-w-tabs-target="panel">
+ *           tab-2
+ *       </section>
+ *   </div>
+ * </div>
+ */
+
+export class TabsController extends Controller<HTMLDivElement> {
+  static targets = ['list', 'trigger', 'label', 'panel'];
+
+  static classes = ['animate'];
+
+  static values = {
+    transition: { default: 150, type: Number },
+    active: { default: '', type: String },
+    disable: { default: false, type: Boolean },
+    animate: { default: true, type: Boolean },
+  };
+
+  declare animateClass: string;
```

Comment:
```suggestion
  declare readonly animateClass: string;
```

Only values can be modified, it's best to change all declare to be readonly to ensure TS can enforce this.